accesshandler/openapi.yml

```diff
@@ -151,6 +151,27 @@ paths:
         name: providerId
         in: path
         required: true
+  "/api/v1/providers/{providerid}/validate":
+    post:
+      summary: ""
+      operationId: validate-request-to-provider
+      responses:
+        "200":
+          $ref: "#/components/responses/HealthResponse"
```

Comment:
Instead of returning `HealthResponse`, can we return a more informative response here? I'm thinking we can reuse some of the ideas from the Diagnostics responses we wrote for the interactive provider setup flow, including how we have a function which returns validations with different IDs:

```go
ValidateConfig() map[string]providers.ConfigValidationStep 
```

This is nice because it makes debugging a lot easier - it gives us the opportunity to log the diagnostics, and users can send us the output when validation fails.

e.g. returning a `GrantValidationResponse` as per the below:

```yaml
components:
  GrantValidation:
        title: GrantValidation
        type: object
        description: A validation of a Grant, without actually creating it.
        properties:
          id:
            type: string
            description: "The ID of the validation, such as `user-exists-in-sso`."
          logs:
            type: array
            items:
              $ref: "#/components/schemas/DiagnosticLog"
        required:
          - id
          - status
          - fieldsValidated
      DiagnosticLog:
        title: DiagnosticLog
        type: object
        description: A log entry related to a validation.
        properties:
          level:
            type: string
            description: The log level.
            enum:
              - INFO
              - WARNING
              - ERROR
          msg:
            type: string
            description: The log message.
        required:
          - level
          - msg
responses:
   GrantValidationResponse:
        description: A validation of a Grant to verify whether it can be created.
        content:
          application/json:
            schema:
              type: object
              properties:
                validation:
                  $ref: "#/components/schemas/GrantValidation"
              required:
                - validation
```

Then, in the Go code, add a `GrantValidationStep` type, similar to how `ConfigValidationStep` works. Note that we'll need to update the `Validator` interface too - we can reuse the same approach we have taken with `ConfigValidator`:
```diff

+type GrantValidationStep struct {
+	Name            string
+	FieldsValidated []string
+	Run             func(ctx context.Context, subject string, args []byte) diagnostics.Logs
+}


// Validators know how to validate access without actually granting it.
type Validator interface {
	// Validate arguments and a subject for access without actually granting it.
-	Validate(ctx context.Context, subject string, args []byte) error
+	ValidateGrant() map[string] GrantValidationStep
}
```

And then an actual validation function would look something like:

```go
func (p *Provider) ValidateGrant() map[string]GrantValidationStep {
  return map[string]GrantValidationStep{
    "user-exists-in-aws-sso": {
      Name: "The user must exist in the AWS SSO instance",
      Run: func(ctx context.Context, subject string, args []byte) diagnostics.Logs {
        // i've left the actual logic out just for brevity
        err := checkExistsInAWSSSO()
        if err != nil {
          return diagnostics.Error(err)
        }
      }
    }
  }
}
```

Reply:
This removes the need to do the concurrency stuff inside the Validate method like we are currently doing - instead, we can run it in the API handler, the same way that ValidateConfig works.